### ISSUE TITLE
pull down to refresh does not work on "INVERTED" FlatList

### ISSUE THREAD

**Issue by quadsurf** [facebook/react-native#17553]
### Is this a bug report?
yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?
yes

### Environment
  Environment: all newer environments
  Packages: any packages
  Target Platform: any platform

### Steps to Reproduce
1. ```return <FlatList {...props} inverted data={[...]} refreshing={this.state.isLoading} onRefresh={this.someListUpdatingFunction} />```
(just add the 'inverted' attribute to any RN ```<FlatList/>``` to reproduce, then try pulling DOWN to refresh, then try pulling UP to refresh)

### Expected Behavior
onRefresh should allow users to refresh using the MOST COMMON way that users have been trained to refresh data over the years, which is to "pull DOWN to refresh" and see an ActivityIndicator spinning ABOVE the FlatList, NEVER below it... this should be the default EVEN WHEN inverted={true}

### Actual Behavior
user has to pull UP to refresh (with Activity Indicator at the bottom) ... literal vs intuitive: when taken "literally", i can agree that this is expected behavior since it is after all "inverted" ... however, when taken "intuitively", it's not so expected, ESPECIALLY for users... no user has ever been trained to pull up to refresh data... EVER!!!

### Most Applicable Use Case
- a chat app, where users pull down to load more chat history that's appended above, similar to how it's done in iOS Messages app
- UIs like terminals, event logs, chat, etc... where it's common to insert new content from the bottom and load old content when the user scrolls to the top or pulls down to refresh

### Demo
https://snack.expo.io/S1UOyWgSM
(just add the 'inverted' attribute to any RN ```<FlatList/>``` component to reproduce)

**Comment by hramos** [comment]
Looks like this may have been closed too eagerly -- the bot didn't see if v53 or newer was used in your issue description.

I'd still argue it's unclear to mention "newer versions" without being explicit, so I'll leave this closed. Please do re-open and make sure to include a specific version.

**Comment by Draccan** [comment]
Hi, I have reproduced the same problem on V53!

**Comment by hramos** [comment]
This issue is straddling the line between bug report and feature request. I'll therefore mark it as a Good first issue for someone to tackle if they would like to get their feet wet with making contributions to React Native.

**Comment by quadsurf** [comment]
@hramos thank you for keeping this alive! I may be able to tackle this in a few weeks, but rather than changing the current behavior, I would just provide other developers with the option to choose for themselves which behavior they prefer between [pull **up** to refresh] or [pull **down** to refresh], while in ```inverted``` mode of course.

**Comment by amloelxer** [comment]
@hramos I would love to work on this as my first issue, but is there a guide for building the React-Native source for iOS? All I see are the steps for Android.

**Comment by hramos** [comment]
I don't think we have a guide for that, but building from source for iOS may not be necessary if the fix only involves JavaScript. 

Open `RNTester/RNTester.xcodeproj` in Xcode and select the RNTester target, and any of the iPhone simulators:
![simulator screen shot - iphone x - 2018-03-14 at 11 58 09](https://user-images.githubusercontent.com/165856/37425468-a923c3ba-2780-11e8-90ba-1d0c81644625.png)

Then open the Flat List Example:

![simulator screen shot - iphone x - 2018-03-14 at 11 58 09](https://user-images.githubusercontent.com/165856/37425485-bb7a147e-2780-11e8-9b0a-49856f85dd92.png)

Confirm everything works as expected. 

You may need to follow the instructions for installing your fork of React Native: http://facebook.github.io/react-native/docs/android-building-from-source.html#1-installing-the-fork

Then you can proceed to add your changes to `Libraries/Lists/FlatList.js` (or SectionList or VirtualizedList, as needed) in your fork, and re-run the Flat List Example to verify. 

Thanks for letting me know these instructions need to be expanded! Let me know how it goes and we can work on improving the docs.

**Comment by amloelxer** [comment]
@hramos will do! Thanks for the info

**Comment by himanshuchauhan** [comment]
If this has not been fixed .I would like to start working on this. Can someone confirm?

**Comment by amloelxer** [comment]
@himanshuchauhan feel free. I spent an hour or two looking at it, but it seemed to me I couldn't do it with just Javascript, and then work and life happened. Curious to your solution if you end up doing it!

**Comment by himanshuchauhan** [comment]
I have been trying to fix this . But I am having issues running the `RNTester/RNTester.xcodeproj` .
I am getting this - 
> no script url provided make sure packager is running or you have embedded a js bundle

**Comment by briank621** [comment]
If anybody else isn't, I'll be taking a look at this

**Comment by briank621** [comment]
I don't think this can be solved with a purely Javascript implementation. The inverted attribute causes a scale transform of -1 on the ScrollView (which is then undone by each cell Item inside to make the cell appear upright). Since the activity indicator would normally appear at the top, it now appears below.

The activity indicator is being generated by the RefreshControl component attached to the ScrollView. A fix that I have in mind is changing the location of the frame of the RefreshControl, as discussed [here](https://stackoverflow.com/questions/17309268/offsetting-uirefreshcontrol?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa), but I haven't had success doing that. From there we would have to change how RefreshControl triggers a refresh (which normally occurs when `scrollY: 0`), so that it happens when scrollY matches the height of the page.

Somebody else can take another try at this. There might be a simpler approach, but the main issue is from the scale transform of -1.

**Comment by endiny** [comment]
Is it really a bug at all? Inverted could mean an inverted refresh either

**Comment by amloelxer** [comment]
@briank621 I thought the exact the same thing. I didn't spend a ton of time looking at it, but remember at the time I didn't think it could be done solely in JS without doing some hackey things. I'll try to take a peak again and maybe build React-Native from source when I have time.

**Comment by giantss** [comment]
@quadsurf  @hramos I am now also making a chat interface, I need to do a chat history load, really need this refresh reversal function.

**Comment by hramos** [comment]
@giantss feel free to send a PR. This issue was marked as a "Good first issue" and it's up to the community to implement.

**Comment by holmeshe** [comment]
I was looking at the issue while I can not see a feature in it. Suggest closing.

In my understanding, there are generally two actions that can operate on a list view, 1) loading more (older) data, and 2) refresh the potential (newer) data.

For normal list view,
![normal](https://user-images.githubusercontent.com/30769400/41693133-046703e8-7558-11e8-8e42-3a4679260932.png)

the edge is on top, so a user has to pull down to refresh (2), and scroll up to load more (1). I think this is because the list of this sort is more common, users feels more natural and intuitive for "pull down to refresh".

When a list is inverted,
![inverted](https://user-images.githubusercontent.com/30769400/41693205-4ebfcc54-7558-11e8-9201-76d5e469c5b9.png)

the edge is on the bottom, so not only the data is inverted, but also the mentioned two operations. And it is the most logical approach because if we cling to "pull down to refresh", then semantically there is no way to load older data (2).

Back to the issue description:

>Most Applicable Use Case
>a chat app, where users pull down to load more chat history that's appended above, similar to how it's done in iOS Messages app --------a)
>UIs like terminals, event logs, chat, etc... where it's common to insert new content from the bottom and load old content when the user scrolls to the top or pulls down to refresh --------b)
```

a) "pull down to load more chat history" is the operation 1). In the `RNTester`, the `FlatList` works **exactly this way**;
b) "...load old content when the user scrolls to the top or pulls down to refresh". I think it's better not to use two separate gestures for "load old content" 1). Because it occupies the gesture for "insert new content". If the "scroll up to refresh" does not feel natural, I would suggest entailing an update interval for "insert new content from the bottom", which is natural as it feels like `tail -f`.

**Comment by mrozbarry** [comment]
I just ran into this, and would like to flush out my use case, and comparison.

First, on iOS, we have a FlatList that used an onScroll hack to load previous conversations with `inverted={true}`. Since iOS's behaviour allows bounce scroll, this worked, and we could trigger our load history button.

This was broken on android, and we opted to fix this using a refresh control. In doing this, and removing our hack, now both iOS and Android require a pull from the bottom to top to trigger the refresh. This seems counter-intuitive, and I've never heard of any chat app requiring a pull up to load history.

The big issue here is the expected flow of a chat app is that recent/latest messages appear at the bottom, which inverted does for us, but it also inverts the refresh direction, which is unexpected.

I think there is one of two solutions: either let RefreshView or FlatList control the direction of the pull to trigger the Refresh, or set a prop on FlatList that specifies the vertical alignment of the items (default to top, but with options for center and bottom).

**Comment by mrozbarry** [comment]
I was able to fix this by wrapping the FlatList in a ScrollView, kept the FlatList inverted, and put the refreshcontrol on ScrollView.

**Comment by dozoisch** [comment]
I personally think the behavior of the RefreshControl being at the bottom makes sense. When you are in a Chat application, where new messages appear at the bottom and you keep the user scrolled at the bottom, it feels natural to pull from the bottom to up to try to refresh the conversation. Slack implements it like that. Having a RefreshControl at the top is impossible due to scrolling top causing loading more messages.

**Comment by ChenLi0830** [comment]
Inspired by @mrozbarry, here is what I did as a workaround. 

```
        <ScrollView
          contentContainerStyle={{
            flexDirection: 'row',
            alignSelf: 'flex-end',
            flexGrow: 1
          }}
          refreshControl={<RefreshControl refreshing={fetchingMore} onRefresh={onFetchMore} />}
        >
          <FlatList
            inverted
            style={styles.list}
            data={messages}
            keyExtractor={this.extractItemKey}
            renderItem={this.renderItem}
          />
        </ScrollView>
```
Note that it's important to have the `flexGrow: 1` style in scrollView, otherwise the FlatList would flow at the top of the scrollview when it's not full.

**Comment by MingyuJeon** [comment]
@ChenLi0830 It makes sense for most chat apps like what you said, However, if the flatList is for showing restaurant list or whatever similar things, then it doesn't make sense.
In this case, RefreshControl should be in the bottom. 
The more fetching items, the more difficult to scrolling up to refresh in the situation user fetched lots of items already and looking at items in the bottom.

**Comment by kiranjd** [comment]
Instruction for installing the forked version in [this](https://github.com/facebook/react-native/issues/17553#issuecomment-373142252) comment. Could someone direct me where I can find it?

**Comment by fabriziobertoglio1987** [comment]
>users have been trained to refresh data over the years, which is to "pull down to refresh" and see an ActivityIndicator spinning above the FlatList, never below it.

In WhatsApp **Pull Down** displays the older chat history.
<p align="center">
  <img src="https://user-images.githubusercontent.com/24992535/79585012-7a7fc180-80cf-11ea-9dfd-bd47d3cccfd9.gif" width="300px"/>
  <img src="https://user-images.githubusercontent.com/24992535/79584989-718ef000-80cf-11ea-8562-723169beb558.gif" width="300px"/>
</p>

In WhatsApp **Pull Up** triggers the refresh:
<p align="center">
  <img src="https://user-images.githubusercontent.com/24992535/79584245-730be880-80ce-11ea-9e30-4889849dce98.gif" width="300px"/>
  <img src="https://user-images.githubusercontent.com/24992535/79584251-743d1580-80ce-11ea-9aa2-daeca616b702.gif" width="300px"/>
</p>

I agree that the `ActivityIndicator` should show on top and never on the bottom and I will try to implement this solution, please let me know if I misunderstood the requirement. Thanks a lot.

**Comment by fabriziobertoglio1987** [comment]
`RefreshControl` prop [`progressViewOffset`](https://reactnative.dev/docs/refreshcontrol#progressviewoffset) works only on `Android` https://github.com/facebook/react-native/issues/10718.

`progressViewOffset` allows to change the position of the `RefreshControl` in a `FlatList` on `Android`.

```javascript


export default function App() {
  const [refreshing, setRefreshing] = useState(false);  
  const screenHeight = Dimensions.get('window').height;
  const onRefresh = () => {
    setRefreshing(true)
  }

  return (
    <FlatList
      data={POSTS}
      renderItem={({ item }) => <Item data={item} />}
      keyExtractor={item => item.id}
      refreshControl={<RefreshControl progressViewOffset={screenHeight - 100} height={100} refreshing={refreshing} onRefresh={onRefresh} />}
      inverted
    />
  );
}
```

You can try the following [snack](https://snack.expo.io/@fabrizio.bertoglio/android-inverted-flatlist) on an `Android` device. The `RefreshControl` will show on top of the screen, as explained in https://github.com/facebook/react-native/issues/10718 it does not work on `iOS`. As I focus on `Android`, I will search for another issue. Thanks a lot :smiley:

**Comment by TheSavior** [comment]
I don't have any context on `progressViewOffset` but it is interesting that that exists. 

If that existed on iOS does that feel like a more proper solution to this problem?

I'd like to avoid adding yet-another-prop to FlatList in https://github.com/facebook/react-native/pull/28857 if there is an existing prop that is intended to enable this behavior.

**Comment by fabriziobertoglio1987** [comment]
>If that existed on iOS does that feel like a more proper solution to this problem?

Yes, but it does not work on iOS with the changes to `RCTScrollView` from https://github.com/facebook/react-native/commit/1d22f8fb27e9432e357401ce6f6cede4d710472c, as Pull Request https://github.com/facebook/react-native/pull/11356 never made it to master.

iOS Developer should check the `diff` and comments from [`RCTRefreshControl` 11356](https://github.com/facebook/react-native/pull/11356/files#diff-cfafa8d7e42ab87ab863f917e187a4ae) and publish a new pull request.

Currently I am working on a pull request for https://github.com/facebook/react-native/issues/18266#issuecomment-624123006 and CircleCI, so I would avoid working with iOS. Thanks a lot :smiley: I wish you a good weekend :smiley:

**Comment by awalmubarak** [comment]
> If that existed on iOS does that feel like a more proper solution to this problem?

I thought the same but it only works on android and not iOS

**Comment by fabriziobertoglio1987** [comment]
> It just moves the refresh control up but you still have to swipe from bottom to top in order to trigger refreshing instead of dragging from top to bottom.

**Comment by fabriziobertoglio1987** [comment]
changes included in https://github.com/facebook/react-native/commit/1d22f8fb27e9432e357401ce6f6cede4d710472c broke `progressViewOffset` on iOS

>this breaks setting negative offsets, I'm using this with contentInset + RefreshControl to make sure it is positioned properly under a overlapping header. This could probably be solved by supporting progressViewOffset on iOS, not sure if there are other use cases for negative offsets. 

https://github.com/facebook/react-native/pull/11356 was meant to fix iOS, but never made it to master

I'll be happy to implement changes for Android if they will be merged in master.

**Comment by sabinayakc** [comment]
Any update on this issue?

**Comment by Dahan-Schuster** [comment]
Just ran into this issue in RN 0.62.2. I believe @ChenLi0830 's workaround isn't the best solution due the Nested VirtualizedList warning ('VirtualizedLists should never be nested inside plain ScrollViews')

**Comment by utkarsh-dixit** [comment]
Is some working on this? If not, I'd like to work on this.

**Comment by acrazing** [comment]
Aha, I solved this problem with the following code:

```ts

        <FlatList
          ref={this.messageList}
          style={styles.list}
          contentContainerStyle={styles.listContent}
          data={messages}
          renderItem={renderItem}
          inverted={true}
          keyExtractor={(item) => item.id.toString()}
          onEndReached={this.fetchMoreMessages}
          ListFooterComponent={<ActivityIndicator />}
          ListFooterComponentStyle={{ flexGrow: 1, paddingTop: 20 }}
        />
```

**Comment by icecapp** [comment]
IMO refresh is, as its name indicates, intended for fresh, new data. It doesn't make sense to use this for past data.
UX-wise, users today are accustomed to infinite scroll when scrolling in the past. Not pull-to-refresh. 
Notice e.g. most chats when reaching the tail of the list use infinite scroll. 
Also if the list and everything in it are inverted, I think it would be really counter-intuitive to not invert the RefreshControl by default, since lists load and stay scrolled at the head, which is never understood the be "the beginning", but rather, "most recent". 
For use cases where this is still necessary, IMO this should really not interfere with `inverted` but rather as an additional prop, e.g. akin to `invertedStickyHeaders` in `SectionList`, you'd get 'invertedRefresh`.

**Comment by RominHalltari** [comment]
@hramos I think @icecapp's argument makes sense. I would like to try to tackle this but I don't know in which direction I should go at this point. Should I introduce the `invertedRefresh` prop and its functionality?

**Comment by AbdelhamidLarachi** [comment]
i solved it by using 
`contentContainerStyle={{ flex: 1, justifyContent: 'flex-end' }}` instead of `inverted`. This will keep displaying your messages from the bottom and keep your RefreshControl on top.

```
<FlatList
refreshControl={<RefreshControl refreshing={refreshing} onRefresh={onRefresh}/>}
contentContainerStyle={{ flex: 1, justifyContent: 'flex-end' }}
renderItem={Message}
data={loadedMessages ? getConversation(conversation, conversations) : MessagesLoaderMockData}
style={styles.MessagesFlatList} />
```

**Comment by marco-hic** [comment]
> i solved it by using `contentContainerStyle={{ flex: 1, justifyContent: 'flex-end' }}` instead of `inverted`. This will keep displaying your messages from the bottom and keep your RefreshControl on top.
> 
> ```
> <FlatList
> refreshControl={<RefreshControl refreshing={refreshing} onRefresh={onRefresh}/>}
> contentContainerStyle={{ flex: 1, justifyContent: 'flex-end' }}
> renderItem={Message}
> data={loadedMessages ? getConversation(conversation, conversations) : MessagesLoaderMockData}
> style={styles.MessagesFlatList} />
> ```

I've used flexGrow instead of flex, flex blocks scrolling :(

**Comment by MariuzM** [comment]
Damn its 4 years and still no native solution from facebook...

**Comment by arjun1194** [comment]
What is the status of this issue ? labels added mention it as a `good-first-issue` and `help-wanted` is it available to be picked up ?

**Comment by fabOnReact** [comment]
https://github.com/facebook/react-native/issues/17553#issuecomment-626225029

**Comment by cyberex7** [comment]
If this issue is open can I take it up?

**Comment by ai-dev-official** [comment]
Is this issue open? If yes, can I pick this up? Are there any specific guidelines or community I should follow?

**Comment by Navjot1806** [comment]
I've submitted a PR to address the visual indicator positioning: #55464

The fix automatically sets `progressViewOffset` on the `RefreshControl` when the list is inverted, repositioning the refresh indicator from the visual bottom to the visual top. The offset is calculated from the visible height of the scroll view after layout.

User-provided `progressViewOffset` values are respected when explicitly set.

Note: The pull-to-refresh gesture direction (tied to native scroll view behavior) would require native-level changes for a complete fix.

**Comment by IkyssOffc** [comment]
I'm interested in working on this. I've been looking into how inverted prop affects the RefreshControl positioning and would like to investigate a fix so that 'pull down' remains the default behavior even when inverted. Please assign this to me.

**Comment by LAKSHYA9941** [comment]
Tested on React Native 0.86.x via Expo SDK 57.

what i Observed:
Pulling down triggers refresh.
- When `inverted={true}`, the refresh indicator is rendered at the bottom of the list.
- When `inverted={false}`, the refresh indicator is rendered at the top.

From testing, the behavior appears consistent with the inverted list semantics.
Hope that helps!!
